### ISSUE TITLE
Electrum daemon hangs when >15000 addresses in wallet

### ISSUE THREAD

**Issue by jwx888** [spesmilo/electrum#6030]
I am running Electrum for accepting Bitcoin payments on a site. Using `createnewaddress` for each customer and then check the addresses with `getaddressbalance` and `getaddresshistory`.

Everything works perfectly until the wallet contains more than ~15000 addresses.

> $ electrum listaddresses | electrum getprivatekeys - | wc -l
> 18576

Daemon randomly hangs on API requests and stops responding. The only way to fix that is to wipe unused addresses in the site's database and start a new wallet.

<sub><sup>
Electrum version: 3.3.8 - https://electrum.org - https://github.com/spesmilo/electrum
Python version: 3.6.8 (default, Jun 28 2018, 19:01:02) 
[GCC 4.8.4]. On platform: Linux-4.9.0-12-amd64-x86_64-with-debian-9.12
</sup></sub>

**Comment by gits7r** [comment]
It would be helpful to know if the CLI daemon would also hang if fit has more than 15000 addresses but properly used in their respective order.

Can you try with by simulating (on testnet maybe) a test with a wallet where you use with electrum on your website more than 15000 addresses but all used in their respective order? Is there an easy way you can automate this and make it?

I helped some people implement electrum payments for their websites but not by using createnewaddress, because I did not want to mess up with electrum's gap limits. Instead, I recommended them to use addrequest with an expiring parameter and a cronjob from time to time manually removes requests long passed their expiration times to free up unused receiving addresses so there's no large portion of unused addresses between used addresses.

**Comment by canihojr** [comment]
wallet.change_gap_limit(15000)

**Comment by jwx888** [comment]
> wallet.change_gap_limit(15000)

I believe that doesn't apply in the context of this issue, as I am using `createnewaddress` which purpose is to "**Create a new receiving address, beyond the gap limit of the wallet**"

Changing the gap limit is only needed when you are using `getunusedaddress`, however I am sure it would result into same issues I am experiencing.


> It would be helpful to know if the CLI daemon would also hang if fit has more than 15000 addresses but properly used in their respective order.
> 
> Can you try with by simulating (on testnet maybe) a test with a wallet where you use with electrum on your website more than 15000 addresses but all used in their respective order? Is there an easy way you can automate this and make it?

I am able to automate or perform almost any test, but can you please elaborate on "using them in their respective order" ? I didn't understand what you mean by this.

**Comment by SomberNight** [comment]
> Daemon randomly hangs on API requests and stops responding. The wallet gets corrupt after you force close unresponsive Electrum.

What do you mean the wallet gets corrupted? Can the wallet file no longer be opened with Electrum?

**Comment by jwx888** [comment]
> What do you mean the wallet gets corrupted? Can the wallet file no longer be opened with Electrum?

I had 2 subsequent issues and wanted to open a separate issue after resolving this one, but since you asked I will respond there:

(1) Some minutes before the wallet was incorrectly closed by rebooting a server, there was a pending API request that was broadcasting an outgoing transaction. The transaction was broadcasted, but apparently has failed adding itself to the wallet history, because it was made during the mentioned 'hang'.  I have found this out after a reboot when the wallet was basically messed up, because the balance was in place, but everytime I tried to make a new 'payto' request, there was an error telling me that electrum couldn't find any available outputs (I don't remember the exact error, sorry). I have fixed this issue by using `addtransaction` to manually add the missing outgoing transaction.

(2) Somewhere in the middle of the previous issue I had a spontaneous desire to try restoring the wallet _using a seed_ in a different fresh instance. When I provided a seed on a new instance, only 10% of funds were restored from seed. I've assumed it's a normal behavior when using new addresses beyond the gap limit and wasn't very concerned at the moment, however I wanted to ask about it after anyway.

**Comment by SomberNight** [comment]
> (2) Somewhere in the middle of the previous issue I had a spontaneous desire to try restoring the wallet using a seed in a different fresh instance. When I provided a seed on a new instance, only 10% of funds were restored from seed. I've assumed it's a normal behavior when using new addresses beyond the gap limit and wasn't very concerned at the moment, however I wanted to ask about it after anyway.

Yes, this sounds normal due to manually adding new addresses. You would need to increase the gap limit (or add addresses manually again) to find all the coins and history.
